### PR TITLE
MGDOBR-872: Manage masked values inside action/source configuration

### PR DESCRIPTION
**Pull request by kelvah**
JIRA: https://issues.redhat.com/browse/MGDOBR-872

This PR depends on:

- https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/863

I'm adding support for the new format of masked values (empty object, `{}`) inside processors.

To test this you will have to check your browser developer tools console and inspect network requests.

Steps to test it:
- `npm run start:mocked` (until the related PR is merged, you can only use the mocked APIs)
- open a bridge and click on create new processor
- select the sink type
- provide a processor name
- select an action with sensitive fields (like `AWS Lambda`)
- fill up the action configuration parameters
- click on the create processor button
- inspect the payload of the POST request that creates the processor. It will include the sensitive fields values as you provided them
- in the processors list page, wait until the new processor gets `ready` and then click on its name
- as the processor detail page opens, inspect the request to `GET /bridges/{bridgeId}/processors/{processorId}`, its response will report an empty object (`{}`) as value for the sensitive fields
- the sensitive values are displayed as a bunch of asterisks in the properties list
- now click on the edit button
- change an optional param like the transformation template
- notice that the sensitive fields in the form are empty, no value is displayed
- save the processor
- inspect the payload of the PUT request that saves the processor; the secret fields values are empty objects (`{}`), meaning the back-end will not overwrite the existing value provided when creating the processor
- finally, go back to the processors list and wait until the processor becomes `ready`
- click on the processor name to open its detail
- click on edit
- change one of the secret parameters by adding a new value
- save the processor
- inspect the payload of the PUT request that saves the processor; the edited secret parameters will have the new value, meaning the back-end will overwrite the previous one.

### CODE REVIEW

File: src/utils/processorUtils.ts
Comment (manstis) [inline src/utils/processorUtils.ts:15]: Is this worthy of more test coverage?

File: src/utils/processorUtils.ts
Comment (manstis) [inline src/utils/processorUtils.ts:72]: It's nice if you point out code, unrelated to this change, that you've moved :smile:

File: src/utils/processorUtils.ts
Comment (manstis) [inline src/utils/processorUtils.ts:116]: I could not see where this is used?

File: src/utils/processorUtils.ts
Comment (manstis) [inline src/utils/processorUtils.ts:169]: This should be worthy of some tests?

**Review by manstis — COMMENTED** [review]
I could not edit a Processor after it had been created.

All _secret_ fields contained the text "This field is a credential and its value cannot be displayed..." however the applicable `TextBox` was disabled too.

**Comment by kelvah** [comment]
> I could not edit a Processor after it had been created.
> All _secret_ fields contained the text "This field is a credential and its value cannot be displayed..." however the applicable `TextBox` was disabled too.

ATM the editing of action/sources parameters is disabled by the UI. I know it's now supported by the APIs but I was unsure if this PR was the right place to enable it.

I guess it is given it will help us evaluate the proper functioning of masked values editing too.
I'm going to do enable it now. Let me know if you have any objections.

**Comment by vpellegrino** [comment]
@kelvah could you please rebase this branch?

**Review by jomarko — COMMENTED** [review]
Thank you for a PR. I removed myself from reviewers to align the state with jira, where @ChughShilpa is a QE Contact. @ChughShilpa please let us know if that is ok. if you have enough time or if you need some help. Thank you.

I agree with @manstis that `processorUtils.ts` should be covered by some tests.

File: src/utils/processorUtils.ts
Comment (kelvah) [reply src/utils/processorUtils.ts:15]: Added some test coverage

File: src/utils/processorUtils.ts
Comment (kelvah) [reply src/utils/processorUtils.ts]: I'll add a review with some comments later to explain the content of `processorUtils`

File: src/utils/processorUtils.ts
Comment (kelvah) [reply src/utils/processorUtils.ts:169]: Done too.

File: src/utils/processorUtils.ts
Comment (kelvah) [reply src/utils/processorUtils.ts:116]: Inside `ProcessorEdit.tsx` :)

File: src/app/Processor/ProcessorDetail/ProcessorDetailConfigParameters.tsx
Comment (kelvah) [reply src/app/Processor/ProcessorDetail/ProcessorDetailConfigParameters.tsx]: I wasn't sure where to put this. In the end I decided to export a const with the asterisks within the same file.

File: src/app/Processor/ProcessorEdit/ConfigurationEdit/ConfigurationEdit.tsx
Comment (kelvah) [inline src/app/Processor/ProcessorEdit/ConfigurationEdit/ConfigurationEdit.tsx:113]: I've moved the retrieval of the schema up one level inside `ProcessorEdit` because I needed it there too.

File: src/app/Processor/ProcessorEdit/ProcessorEdit.tsx
Comment (kelvah) [inline src/app/Processor/ProcessorEdit/ProcessorEdit.tsx:139]: Moving `prepareFilters` and `prepareRequest` functions to `processorUtils`. 

They are used to convert the data coming from the processor form model into the payload for `ProcessorRequest`.

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:15]: This is responsible for handling secret parameters values.
Used in particular when passing initial data to the generated form and when preparing data to be sent to processor requests.

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:72]: Moved from `ProcessorEdit.tsx` to keep all the helpers in one place.

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:116]: Moved from `ProcessorEdit.tsx`

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:169]: Retrieving the definition of a property inside a schema. Used when handling secret parameters to decide if the parameter is defined as secret or not.

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:206]: Basically a type guard to speed up schema inspection.

File: src/utils/processorUtils.ts
Comment (kelvah) [inline src/utils/processorUtils.ts:218]: Really basic reference resolver to retrieve the definition of a property specified in a sub-schema. Only supports basic sub-schemas defined in the same file. (They can be also relative or absolute links to other files)

File: src/app/Processor/CreateProcessorPage/CreateProcessorPage.tsx
Comment (kelvah) [inline src/app/Processor/CreateProcessorPage/CreateProcessorPage.tsx:207]: Moved this outside `ProcessorEdit` because in theory it should be a standalone component, not containing logic tied to specific page URLs.

File: src/app/Processor/ProcessorEdit/ConfigurationEdit/ConfigurationEdit.tsx
Comment (kelvah) [inline src/app/Processor/ProcessorEdit/ConfigurationEdit/ConfigurationEdit.tsx:155]: We need to know if we are creating a new processor or editing an existing one. The handling of secret values will depend on it.

**Review by kelvah — COMMENTED** [review]
I've updated the behavior of the processor form to align it with what the APIs allow currently.

So it is **not** possible to edit a processor name, while **it is** now possible to edit the action/source configuration parameters of a processor.

**Review by manstis — COMMENTED** [review]
A couple of points I noticed (before I spend any time reviewing the code):
- If I edit a Processor and save it, the "Edit" button is disabled after the save. If I want to edit it again I need to first navigate back to the Instance and then select the Processor again.
- When creating a Processor the _password_ field (e.g. Slack Token) is showing asterisks (i.e. a "password" textbox) as I type. Is this by design? I was expect to be able to see what I was typing to be sure I hadn't made any mistakes.

**Comment by kelvah** [comment]
> A couple of points I noticed (before I spend any time reviewing the code):
> 
> * If I edit a Processor and save it, the "Edit" button is disabled after the save. If I want to edit it again I need to first navigate back to the Instance and then select the Processor again.

That's because you have to wait until the processor gets into the `ready` state again before being able to edit it. 

The detail page of a processor is not automatically refreshed during provisioning. What you did is the easiest way to go back to editing it without having to manually reload the detail page, which could be slower.

The current behavior predates this PR by the way.
We discussed about possible improvements in the past; like adding a "refresh" button to the processor detail page. Still a bit unsure about it. 

> * When creating a Processor the _password_ field (e.g. Webhook URL) is showing asterisks (i.e. a "password" textbox) as I type. Is this by design? I was expect to be able to see what I was typing to be sure I hadn't made any mistakes.

I know, It isn't great. I don't like it either. 
That's how password fields are managed by our libraries/setup, using a [`<TextInput />`](https://www.patternfly.org/v4/components/text-input) with `type="password"`.
It doesn't make much sense for a Webhook URL field. It could be acceptable for a "secret key" or similar. 
Unfortunately we have only one type of "secret" field, so they are all treated the same way.

We could add at least a button to show/hide the content of the field next to it.
We can use the `<InputGroup />` component like it's done in [this login form example](https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/LoginPage/LoginForm.tsx#L126).
This of course needs to be implemented in the `uniforms-patternfly` library, we are not directly choosing which component is used to render a particular field.

I'll bring this topic to the next UX sync. I agree we have to do something.

**Review by manstis — APPROVED** [review]
Thank-you for the improvements @kelvah :+1:

**Review by vpellegrino — APPROVED** [review]
No comment from my side. Thank you @kelvah

**Comment by sjamboro** [comment]
> Thank you for a PR. I removed myself from reviewers to align the state with jira, where @ChughShilpa is a QE Contact. @ChughShilpa please let us know if that is ok. if you have enough time or if you need some help. Thank you.
> 
> I agree with @manstis that `processorUtils.ts` should be covered by some tests.

Hi @jomarko and @ChughShilpa I have missed this conversation. I have created MGDOBR-967 to handle tests which are connected with Actions.

**Review by ChughShilpa — COMMENTED** [review]
Hi @kelvah 
I found masking is not present for  parameters (basic_auth_username , basic_auth_password) with format: "password" for Sink processor with Webhook action type

**Review by ChughShilpa — COMMENTED** [review]
Another Observation, in some cases when editing processor, sensitive fields are not empty (value is already present in asterisks) for below cases:
1. slack_source  >> Token field  
2. aws_lambda_sink  >> Access key
3. slack_sink >> webhookurl field 

and for "slack_sink" - while editing the processor, add a transformation template and then save, it shows an error 
"for webhook as "must match pattern "(http|https):\/\/([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:\/~+#-]*[\w@?^=%&\/~+#-])"'
but value is already present (which too shouldn't be the case).
However, adding value for webhook and then save the processor - This works

**Review by ChughShilpa — COMMENTED** [review]
One more observation, sensitive fields value is already present when creating a new processor like
1. aws_sqs_source  >> 'access key' field
2. Slack_sink >> 'webhook url' field

Is it by Design?

**Comment by kelvah** [comment]
> Hi @kelvah I found masking is not present for parameters (basic_auth_username , basic_auth_password) with format: "password" for Sink processor with Webhook action type

@ChughShilpa You're right, masking is not properly working for those fields. 
There are some issues with the current schemas. Not all the "secret" fields have been defined properly. The schemas were added at different stages by different people, the correct way of defining secrets was not yet clear by then.

At the moment the secret fields working properly with this PR are those defined with the `oneOf` property and two sub-schemas, like [the slack_webook_url for example](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/blob/main/mocked-api/schemasData.ts#L202). 

The fields with just the `format: "password"` property, like the two you mentioned, need to be fixed.
I saw Alessandro [already fixed](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/863/files#diff-1563220af235a211bc9a6b59a7abbf55c76742091a74bcf6922d6653087f2abc) a bunch of them in [his PR](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/863).  We'll make sure he covered all of them when reviewing it.

**Comment by kelvah** [comment]
> Another Observation, in some cases when editing processor, sensitive fields are not empty (value is already present in asterisks) for below cases:

> One more observation, sensitive fields value is already present when creating a new processor like
> 
> 1. aws_sqs_source  >> 'access key' field
> 2. Slack_sink >> 'webhook url' field
> 
> Is it by Design?

Both things shouldn't happen, but I'm not able to replicate them.

Let's have a chat tomorrow to figure out what's going on.

**Comment by ChughShilpa** [comment]
> > Another Observation, in some cases when editing processor, sensitive fields are not empty (value is already present in asterisks) for below cases:
> 
> > One more observation, sensitive fields value is already present when creating a new processor like
> > 
> > 1. aws_sqs_source  >> 'access key' field
> > 2. Slack_sink >> 'webhook url' field
> > 
> > Is it by Design?
> 
> Both things shouldn't happen, but I'm not able to reproduce them.
> 
> Let's have a chat tomorrow to figure out what's going on.

Discussed with @kelvah and found out that this is typically because of the value of parameter having ( format: "password") saves in automatically by Google (although didn't save the values explicitly) and thus picking the last filled values for the same parameter while creating and editing the processor.

**Comment by kelvah** [comment]
Thanks @ChughShilpa. 
I think the issue only happens when submitting the form and giving permission to store form data when google prompts the user to do so.
I've created [a new ticket](https://issues.redhat.com/browse/MGDOBR-981) to investigate the issue.

**Comment by kelvah** [comment]
Hi @ChughShilpa, the [back-end PR for masked values](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/863) has been merged & deployed.
Waiting for your final review/approval here to proceed with this one.

**Comment by ChughShilpa** [comment]
> Hi @ChughShilpa, the [back-end PR for masked values](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/863) has been merged & deployed. Waiting for your final review/approval here to proceed with this one.

@kelvah  Changes are verified. Working as expected

**Review by ChughShilpa — APPROVED** [review]

**Pull request merged**
